Fix handling of HEAD requests in UrlConnectionHttpClient

abrooksv opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## Motivation and Context
More fixes to support HttpUrlConnection

## Testing
New tests in SdkHttpClientTestSuite, manual test works.

```
        S3Client.builder()
            .build().headBucket {
                it.bucket("IDontExist141425425")
            }

The specified bucket does not exist. (Service: S3, Status Code: 404, Request ID: 85007F24FB97EA0C)
software.amazon.awssdk.services.s3.model.NoSuchBucketException: The specified bucket does not exist. (Service: S3, Status Code: 404, Request ID: 85007F24FB97EA0C)
	at software.amazon.awssdk.services.s3.model.NoSuchBucketException$BuilderImpl.build(NoSuchBucketException.java:117)
```


## Screenshots (if appropriate)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](../docs/LaunchChangelog.md)

## License
<!--- The SDK is released under the Apache 2.0 license (http://aws.amazon.com/apache2.0/), so any code you submit will be released under that license -->
<!--- For substantial contributions, we may ask you to sign a Contributor License Agreement (http://en.wikipedia.org/wiki/Contributor_License_Agreement) -->
<!--- Put an `x` in the below box if you confirm that this request can be released under the Apache 2 license -->
- [x] I confirm that this pull request can be released under the Apache 2 license

kiiadi commented on http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java:
```diff
@@ -140,12 +141,23 @@ private RequestCallable(HttpURLConnection connection, HttpExecuteRequest request
+                sdkHttpMethod == SdkHttpMethod.PUT) {
+
+                // TODO: Does this need to wrap like in ApacheHttpRequestFactory#wrapEntity?
+                request.contentStreamProvider().ifPresent(provider ->
```
I wonder if the `contentStreamProvider` should actually return null if it's an unsupported `SdkHttpMethod` - otherwise every implementation of an HTTP client has to have this logic.

abrooksv replied:
It seems it does, the test harness always passed it, so fixed the test harness

abrooksv commented:
Updated PR, the SDK actually does provide a null content stream provider when no body but the test harness always did. 
Updated test harness to fix that and thus the production change is smaller
